### ISSUE TITLE
initialRegion not correctly handled.

### ISSUE THREAD

**Issue by heysailor** [react-community/react-native-maps#1617]
React Native Maps 0.16.3, React Native 0.47, Google maps on Android and iOS.

Unfortunately, I'm finding that the initialRegion prop of MapView is still ignored in both android and iOS with Google maps. 

On Android, I get a whole-world view, with latitude and longitude of 0. On iOS, the same latitude and longitude, but zoomed right in.

I have verified that my provided initialRegion is valid. 

Appreciate everyone's work.

**Comment by heysailor** [comment]
This is likely the same as discussed in #1559 - but I'm finding it on iOS as well. Using region over initialRegion works.

**Comment by heysailor** [comment]
Looking into this a bit more - logging out `initialRegion` from `MapView._onReady` and `MapView._onLayout`, and `event.nativeEvent.region` from `MapView._onChange`.

I'm not very familiar with the guts of react-native-maps, so if anyone can chime in and give pointers, that'd be helpful.

**Google Maps on iOS:**

_On first load:_
`MapView._onReady` logs a correct `initialRegion`
`MapView._onLayout` does not appear to fire. 
--> `initialRegion` is honoured and the map displays correctly.

_When the map component is unmounted then re-mounted:_
`MapView._onLayout` is fired with the correct `initialRegion`.
`MapView._onChange` fires followed by and logs a modified `region` which includes longitude of zero and latitude of essentially zero.
`MapView._onReady` fires and logs a modified `initialRegion` which is identical to the `region` of `onChange`.
--> `initialRegion` is not honoured, instead the map shows the blue screen of just sea at this point of the earth.

**Google Maps on android:**

_On first load:_
`MapView._onReady` logs a correct `initialRegion`
`MapView._onLayout` does not appear to fire
`MapView._onChange` logs a `region` with latitude and longitude of zero
--> The map reflects that `initialRegion` has been ignored.

In my implementation, I store the region in redux after user map moves. If I move the map, all is fine and my stored values reflect the map.

_When the map component is unmounted then re-mounted with the updated, post-move `initialRegion` out of redux:_
`MapView._onReady` is fired with the correct `initialRegion`
`MapView._onLayout` is fired with the correct `initialRegion`
`MapView._onChange` fires and logs a modified `region` which includes longitude of zero and latitude of essentially zero. 
(Note `_onReady` fires before `_onLayout`, the reverse of iOS)
--> Again, the `initialRegion` prop is ignored as above and the map reverts back to the world view centred on latitude, longitude of zero.

What is happening to change the map region internally as the map is mounted?

**Comment by heysailor** [comment]
Update: I can partially fix the issue in Google Maps on iOS by commenting out two lines in `ios/AirGoogleMaps/AirGoogleMap.m`:

    - (void)setInitialRegion:(MKCoordinateRegion)initialRegion {
      // if (_initialRegionSet) return;
      // _initialRegionSet = true;
      self.camera = [AIRGoogleMap makeGMSCameraPositionFromMap:self andMKCoordinateRegion:initialRegion];
    }

- partially, because removing this check ensures the` initialRegion` is set and the map displays correctly, but then the user can no longer move the map.

I'm guessing that `_initialRegionSet` is preserved as `true` between unmounting and remounting a component, preventing update of the camera position.

In android, the equivalent is in `AirMapView.java` - but commenting out the checks has no effect:

      public void setInitialRegion(ReadableMap initialRegion) {
        if (!initialRegionSet && initialRegion != null) {
          setRegion(initialRegion);
          initialRegionSet = true;
        }
      } 

If there's someone else who can continue on from here, that'd be great.

As a temporary workaround, I have used the `onMapReady` callback of `MapView`, and call `animateToRegion` to display the `initialRegion`. This means I temporarily get the erroneous display, then the region I want.

**Comment by computerjazz** [comment]
@heysailor did you do a clean build after upgrading the library to 0.16.3? There were some native changes included that won't take effect unless the native modules are rebuilt.

I was seeing the same issues you described but I no longer see them on 0.16.3

**Comment by christopherdro** [comment]
NOTE:

This change was ultimately reverted because the correct fix had been merged but not yet released.

**Comment by omaksi** [comment]
I've just tested 16.3 and it seems #1577 still persists

**Comment by kelset** [comment]
Same issue here, RN 0.46.4 & maps 0.16.4

**Comment by 8-uh** [comment]
Same Issue here: RN 0.48.4 and maps 0.16.4

**Comment by aaronbuchanan** [comment]
Same on (my fork)[https://github.com/aaronbuchanan/react-native-maps]

**Comment by HerbertLim** [comment]
Same issue here: RN 0.48.4, expo 21.0.3, maps 0.15.4 and 0.17.1

As soon as I start my app, most of the time map shows Lat/Lng 0,0. Sometimes, it shows correct initialRegion. Can't tell when it shows correct region and wrong region, (0,0).

I was using react-native-maps in Expo, which was 0.15.4.
So I installed recent maps, and tried to use that but same happens.
(before: import {MapView} from 'expo', now: import MapView from 'react-native-maps')

Since my app's initial loading time is rather long, users have to see the wrong region more than ten seconds.

**Comment by HerbertLim** [comment]
I succeeded avoiding this issue by deferring heavy load jobs to start 1 second after MapScreen component is mounted. 

Here is my guess:
It seems that map is rendered at Lat/Lng 0,0 at first time and quickly change it to initialRegion. In normal case, it is quick enough that we cannot see unwanted region. However, if there are heavy initial tasks from the beginning, the job to change map's region to initialRegion is delayed, which makes us to see unwanted region for a while.

My MapScreen component uses MapView and its componentDidMount() calls an action creator this.props.reload(), which invoke an API to get huge data. I just deferred this one second after cDM like that.

`componentDidMount() {
        setTimeout( () => {
            this.props.reload();
        }, 1000)
    }`

Now, I cannot see unwanted region (0,0) anymore. According to the log, map's initial region is (0,0) but seems to change its region very quickly to my initialRegion.

**Comment by nezaidu** [comment]
I have a little bit different problem. On first component mount, initial region is applied. On second, it is not applied. Anyone experiences the same?

**Comment by alvelig** [comment]
That's fixed already, but let us now if not.

**Comment by SimonErich** [comment]
We are still facing this issue.

Using:
react-native: 0.54.4
react-native-maps: 0.21.0
react-native-maps-super-cluster: 1.4.0

We are user react-native-maps-super-cluster, so maybe it's a problem with that, but it seems odd that it fits the errors above perfectly.

On Android everything is fine, but on ios it seems like, it is zooming to the current region first and then just stops there, ignoring the intiialregion all together.

I've logged and checked everything to make sure the correct region is used for initialRegion. 
We also tried the componentDidMount( requestAnimationFrame, trick, but that didn't work either.

Right now we block the first 3 seconds of regionChanges and have a timeout of 3 seconds after componentDidMount to do the animation to the correct location.
That works, but it doesn't look good at all, because the map keeps jumping around quite a bit.

Like I said: on android everything works perfectly fine.

**Comment by reneeichhorn** [comment]
I'm experiencing the same issue. I'm logging my initialRegion and it has the correct values on first render / when mounting. But it still renders the world map.
I'm using the AnimatedMapView so maybe it's related to that.

**Comment by jackimatin** [comment]
hi 
same issue

**Comment by devendra-asahi** [comment]
I added an initial state of "{}" empty object instead of null , Since is was useState to set Latitude and Longitude using setState

**Comment by RobHedley** [comment]
I know this has been closed but If anyone is still struggling with this issue as I was, I managed to fix it.

In my case I was using animateCamera to rotate the map the match the users heading as they moved. I think in my situation there was a race condition where the compass state would change which would fire the animateCamera function before the Map was fully mounted. This would stop the re-render of the map from its initialisation location of lat: 0 and long:0 leaving the user looking at the middle of the Atlantic Ocean.

I fixed this race condition by adding a ref to the mapView, created an animateMap local state (which I will use in the settings page so the user can turn off the rotation on the map if they want) and then added the following use effect which is fired when the mapView renders:

useEffect(() => {
    if(theMap){
      setAnimateMap(true);
    }
  }, [theMap]);

In essence, the initial state of animateMap is set to false on initialisation. The useEffect is triggered when anything changes with the theMap ref and the state of animateMap is updated to true if the theMap ref exists (it has mounted). The following code then handles the changes from compassHeading and only animates the map if the animateMap state is true.

Hope that helps as I struggles to find a lot of help for this one.
